server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -322,29 +323,39 @@ void executeRequest(
             relativeStartNanos,
             System::nanoTime
         );
+
+        final ClusterState clusterState = clusterService.state();
+        final SearchContextId searchContext;
+        final Map<String, OriginalIndices> remoteClusterIndices; // key to map is clusterAlias
+        if (original.pointInTimeBuilder() != null) {
+            searchContext = original.pointInTimeBuilder().getSearchContextId(namedWriteableRegistry);
+            remoteClusterIndices = getIndicesFromSearchContexts(searchContext, original.indicesOptions());
+        } else {
+            searchContext = null;
+            remoteClusterIndices = remoteClusterService.groupIndices(original.indicesOptions(), original.indices());
+        }
+        final OriginalIndices localIndices = remoteClusterIndices.remove(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
+        final IndicesOptions indicesOptions = original.indicesOptions();
+
+        final AtomicReference<Index[]> resolvedLocalIndices = new AtomicReference<>();
+        final Supplier<Index[]> resolvedLocalIndicesSupplier = () -> {
+            resolvedLocalIndices.compareAndSet(null, resolveLocalIndices(localIndices, clusterState, timeProvider));
+            return resolvedLocalIndices.get();
+        };
```

Comment:
It would be nice if we could just always resolve the local indices here, but unfortunately two problems prevent us from doing that:

- Point in time requests. These requests resolve the indices when the point in time is established and use those resolved indices in all search requests that reference the point in time. When handling such search requests, the concrete indices aren't resolved at all because one or more of the referenced indices could have since been deleted. Instead, `TransportSearchAction` uses the resolved indices from the point in time and allows for partial failures in case of index deletion.
- A small number of query tests expect index resolution to happen after query rewrite logic is executed. Always resolving the indices early breaks those tests.

The first point is the main problem. The previous index resolution approach in `TransportSearchAction` deferred index resolution until after query rewrite logic is executed and made it conditional on a non point-in-time request. This new approach keeps those attributes to prevent regressions, while also creating a resolved local index supplier we can use to lazily generate the index metadata map as necessary.

Reply:
Will add comments, it's definitely non-obvious

Reply:
Added: https://github.com/elastic/elasticsearch/pull/107075/commits/e6274772a90c972eaf937a4a4a5cc9c71ab6fa31